### ISSUE TITLE
Numba Error while running fit_transform with python 3.9

### ISSUE THREAD

**Issue by Lathomas42** [lmcinnes/umap#572]
Using python 3.9 on ubuntu 16.04

I get this issue when running this code:

```
import umap
from sklearn.datasets import load_digits

digits = load_digits()

embedding = umap.UMAP().fit_transform(digits.data)

```
```

TypingError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function make_quicksort_impl.<locals>.run_quicksort at 0x7f9fa1787160>) found for signature:
 
 >>> run_quicksort(array(float64, 1d, C))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'register_jitable.<locals>.wrap.<locals>.ov_wrap': File: numba/core/extending.py: Line 150.
    With argument(s): '(array(float64, 1d, C))':
   Rejected as the implementation raised a specific error:
     UnsupportedError: Failed in nopython mode pipeline (step: analyzing bytecode)
   Use of unsupported opcode (LOAD_ASSERTION_ERROR) found
   
   File "../../../../../fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/numba/misc/quicksort.py", line 180:
       def run_quicksort(A):
           <source elided>
               while high - low >= SMALL_QUICKSORT:
                   assert n < MAX_STACK
                   ^
   
  raised from /auto/fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/numba/core/byteflow.py:269

During: resolving callee type: Function(<function make_quicksort_impl.<locals>.run_quicksort at 0x7f9fa1787160>)
During: typing of call at /auto/fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/numba/np/arrayobj.py (5032)


File "../../../../../fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/numba/np/arrayobj.py", line 5032:
    def array_argsort_impl(arr):
        return sort_func(arr)
        ^

During: lowering "$v.22 = call $54load_attr.5($52binary_subscr.4, func=$54load_attr.5, args=[Var($52binary_subscr.4, utils.py:32)], kws=[('kind', Var($const56.6, utils.py:32))], vararg=None)" at /auto/fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/umap/utils.py (32)
During: lowering "id=0[LoopNest(index_variable = parfor_index.18, range = (0, $38binary_subscr.17, 1))]{46: <ir.Block at /auto/fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/umap/utils.py (30)>}Var(parfor_index.18, utils.py:30)" at /auto/fhome/lthomas/.virtualenvs/py39/lib/python3.9/site-packages/umap/utils.py (30)

```

Versions:
umap-learn (0.5.0)
numpy (1.19.4)
numba (0.51.2)
scipy (1.5.4)
pynndescent (0.5.1)
scikit-learn (0.23.2)

**Comment by lmcinnes** [comment]
Python 3.9 is not supported yet for numba. I believe Numba 0.53, due soon, will fix that, but for now it is a matter of waiting for Python 3.9 support to arrive.

**Comment by leriomaggio** [comment]
Yes exactly @Lathomas42 - Python 3.9 has been integrated into the latest _release candidate_ of `numba`. 
I am currently testing `numba` on Python 3.9 on macOS and Linux (w/ GPU) - if you would have time to replicate those tests on your ends too, it would be wonderful! 

More on issue [#6345](https://github.com/numba/numba/issues/6345) on `numba` repository, if interested!

Besides, in addition to run `numba` test battery by itself, it would be also very useful imho to test `numba 0.53` in combination with `umap 0.5.1` too (as target application for `numba`) 

@lmcinnes (this is an heads up for you for a new PR coming very soon 😄 )
I started working on testing `numba-0.53.0rc1` with `umap`, I ended up optimising the `umap` test battery too.
Mainly: (a) dropping `nose` dependency and moving everything on `pytest` (`nose` not really needed, imho); 
(b) [specifically related to the task at hand] allowing to run the whole `umap` test battery even with incomplete dependencies (e.g. `tensorflow` not installed in the environment).

This was mainly motivated by the fact that apparently there is no intention to extend support to Python 3.9 in `tensorflow` (see [#44485](https://github.com/tensorflow/tensorflow/issues/44485) and this would affect `ParametricUMAP`.

**Comment by leriomaggio** [comment]
Speaking of, I am happy to report that all the tests executable with `Python 3.9` and using `numba==0.53.0rc1` **PASS**

```
================================================================================= test session starts =================================================================================
platform linux -- Python 3.9.1, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/valerio/repos/umap_39
collected 157 items

test_aligned_umap.py ...                                                                                                                                                        [  1%]
test_composite_models.py .s..                                                                                                                                                   [  4%]
test_densmap.py .s.                                                                                                                                                             [  6%]
test_parametric_umap.py ssssss                                                                                                                                                  [ 10%]
test_plot.py .s                                                                                                                                                                 [ 11%]
test_umap_metrics.py ..........................................                                                                                                                 [ 38%]
test_umap_nn.py ..sssssss..                                                                                                                                                     [ 45%]
test_umap_on_iris.py .........                                                                                                                                                  [ 50%]
test_umap_ops.py ......................                                                                                                                                         [ 64%]
test_umap_repeated_data.py .........                                                                                                                                            [ 70%]
test_umap_trustworthiness.py ..........                                                                                                                                         [ 77%]
test_umap_validation_params.py ....................................                                                                                                             [100%]

============================================================== 141 passed, 16 skipped, 65 warnings in 224.84s (0:03:44) ===============================================================

**Comment by jondo** [comment]
I guess this can be closed, now that Numba 0.53 is released.
